dali/python/nvidia/dali/plugin/paddle.py

```diff
@@ -0,0 +1,383 @@
+    return ptr
+
+
+def recursive_length(tensor, lod_level):
+    def _recurse(data, result, level):
+        if level > 0:
+            if isinstance(data, (TensorListCPU, TensorListGPU)):
+                # handle tensor list
+                length = len(data)
+                result[0].append(length)
+                for i in range(length):
+                    _recurse(data.at(i), result[1:], level - 1)
+            elif hasattr(data, 'shape'):
+                # handle dense GPU tensors and numpy.ndarray
+                shape = data.shape
+                if callable(shape):
+                    shape = shape()
+                length = shape[0]
+                result[0].append(length)
+                for i in range(length):
+                    _recurse(shape[1:], result[1:], level - 1)
+            else:
+                # handle shape
+                length = data[0]
+                result[0].append(length)
+                for i in range(length):
+                    _recurse(data[1:], result[1:], level - 1)
+
+    seq_len = [[] for _ in range(lod_level)]
+    _recurse(tensor, seq_len, lod_level)
+    return seq_len
+
+
+class DALIGenericIterator(object):
+    """
+    General DALI iterator for Paddle. It can return any number of
+    outputs from the DALI pipeline in the form of Paddle's Tensors.
+
+    Parameters
+    ----------
+    pipelines : list of nvidia.dali.pipeline.Pipeline
+                List of pipelines to use
+    output_map : list of str or pair of type (str, int)
+                 The strings maps consecutive outputs of DALI pipelines to
+                 user specified name. Outputs will be returned from iterator
+                 as dictionary of those names. Each name should be distinct.
+                 Item can also be a pair of (str, int), where the int value
+                 specifies the LoD level of the resulting LoDTensor.
+    size : int
+           Number of samples in the epoch (Usually the size of the dataset).
+    auto_reset : bool, optional, default = False
+                 Whether the iterator resets itself for the next epoch
+                 or it requires reset() to be called separately.
+    stop_at_epoch : bool, optional, default = False
+                 Whether to return a fraction of a full batch of data
+                 such that the total entries returned by the
+                 iterator == 'size'. Setting this flag to False will
+                 cause the iterator to return the first integer multiple
+                 of self._num_gpus * self.batch_size which exceeds 'size'.
+    dynamic_shape: bool, optional, default = False
+                 Whether the shape of the output of the DALI pipeline can
+                 change during execution. If True, the LoDTensor will be
+                 resized accordingly if the shape of DALI returned tensors
+                 changes during execution.
+                 If False, the iterator will fail in case of change.
```

Comment:
Is there any reason why `fill_last_batch` and `last_batch_padded` is not supported like in MXNet and PyTorch iterators?

Reply:
done, just found out these while adding the tests, the plugins code was actually written before those was introduced.